### ISSUE TITLE
GroupedBar: second group not shown when only one name present

### ISSUE THREAD

**Issue by scottbw** [britecharts/britecharts#824]
I'm using GroupedBar within a StencilJS component. When I render the Grouped Bar and the dataset contains e.g.:

```
[{name: "10006022", group: "main", value: 8888, topicName: "main"}
{name: "10006022", group: "benchmark", value: 8888, topicName: "benchmark"}]
```

... then the rendered chart only shows one of the bars. E.g.:

```
<g class="chart-group">
<g transform="translate(0,18)" class="layer">
<rect class="bar" x="1" y="81" height="66" fill="grey" style="opacity: 1;" width="812"></rect>
<rect class="bar" x="1" y="8" height="66" fill="#f866b9" style="opacity: 0.24;"></rect>
</g>
</g>
```

Note that the second rect element has no width attribute!

In the case above, I would expect the rendered chart to have two bars, in this case of equal length.

Below is my initialisation code for the chart:

```
createHorizontalBarChart() {
      this.tooltip = britechart.tooltip();
      this.barChart = britechart.groupedBar()
        .valueLabel('value')
        .groupLabel('group')
        .nameLabel('name')
        .on('customMouseOver', this.tooltip.show)
        .on('customMouseMove', this.updateTooltip.bind(this))
        .on('customMouseOut', this.tooltip.hide)
      .isHorizontal(true)
      .margin({
        left: 120,
        right: 20,
        top: 20,
        bottom: 30
      })
      .isAnimated(true)
      .colorSchema(this.colorschema)
      .width(this.width)
      .height(this.height);

    this.barChartElement.datum(this.data).call(this.barChart);
    this.tooltipContainer = select(this.element.shadowRoot.querySelectorAll('.metadata-group')[0]);
    this.tooltipContainer.datum([]).call(this.tooltip);
  }
```

## Your Environment
* Britecharts version used: 2.16
* Browser Name and version: Chrome, latest
* Operating System and version: MacOS

**Comment by Golodhros** [comment]
Hi @scottbw, thansk for opening this issue!

It looks like a bug. Could you try to reproduce it on a codepen? You could use this one as a jumpstart: https://codepen.io/Britecharts/pen/MEqZmq

Thanks!

**Comment by scottbw** [comment]
I'm having a lot of difficulty replicating this one, its possible it only occurs in the specific conditions of compiling a component in StencilJS and deploying the chart in a shadow DOM. I've narrowed it down a little to only occurring in GroupedBar when animated=true. It behaves fine when not animated. So something about the 'tween' function doesn't work quite right in my scenario, but only when there are only two rows as in the description above.

**Comment by scottbw** [comment]
Here's a minimal test to reproduce.

Create a new StencilJS project with `npm init stencil`, with the target of creating a Web Component.

Edit the my-component.tsx template as follows:

~~~
import {Component, h, Element} from '@stencil/core';
import britecharts from 'britecharts';
import {select} from 'd3-selection';

@Component({
  tag: 'britecharts-bug-824',
  styleUrl: 'my-component.css',
  shadow: true
})
export class MyComponent {

  @Element() element: HTMLElement;

  componentDidLoad() {
    let barChartElement = select(this.element.shadowRoot.querySelectorAll('.chart')[0]);
    let barChart = britecharts.groupedBar()
      .valueLabel('value')
      .groupLabel('group')
      .nameLabel('name')
      .isAnimated(true);

    let data =
      [{name: "fruit", group: "apple", value: 5},
      {name: "fruit", group: "banana", value: 10}];


    barChartElement.datum(data).call(barChart);
  }
  render() {
    return <div class="chart"></div>;
  }
}
~~~

Compile with `npm run build` and view the output. The chart only shows one bar, the other has no height attribute. Removing `isAnimated(true)` shows both bars rendering correctly.

**Comment by Golodhros** [comment]
Hi @scottbw, thanks for diving on this!

Interesting, so the animation part is what it messes up things! I am not familiar with Stencil, but could it be that the issue has to do with the time it takes to animate it? Maybe at some point, our chart is losing the reference of the element it was tweening over?

**Comment by scottbw** [comment]
I'm struggling to think of a reason why it only fails when there are two bars but is fine with three!

Stencil doesn't do anything particularly odd - it renders the template elements into a shadow DOM within the custom web element, and in this case only invokes Britecharts after those elements are rendered.

**Comment by SazanYa** [comment]
Hello everyone, I have the same problem. I'm using GroupedBar with the following dataset:

```
[
   { group: "Unique Visits", name: "2011-01-05T00:00:00", value: 56 },
   { group: "Leads", name: "2011-01-05T00:00:00", value: 12 },
   { group: "Calls", name: "2011-01-05T00:00:00", value: 6 },
   { group: "Unique Visits", name: "2011-01-06T00:00:00", value: 78 },
   { group: "Leads", name: "2011-01-06T00:00:00", value: 23 },
   { group: "Calls", name: "2011-01-06T00:00:00", value: 5 }
]
```

and it shows only two of the bars, when I expected three bars in each group. The last bar is invisible, because of `height` attribute is absent. If I reduce previous dataset to the following:

```
[
   { group: "Unique Visits", name: "2011-01-05T00:00:00", value: 56 },
   { group: "Leads", name: "2011-01-05T00:00:00", value: 12 },
   { group: "Calls", name: "2011-01-05T00:00:00", value: 6 }
]
```

then only one bar is shown in the rendered chart, the last two are invisible again.

When `isAnimated` setting is setted to `false` then it shows all bars. And when I have three or more groups (add one more record with new `name` property to the first dataset, for example) it shows all bars again. 

The problem occurs when groups count is less than bars count in group (for the first dataset it is 2 and 3 respectively) and `isAnimated` is `true`. Hope it's clear.

I'm not using any frameworks, just vanilla js (sometimes jQuery) and html + css.
Code to reproduce:

```
<div class="bar-container"></div>
```

```
let container = d3.select('.bar-container');
let groupedBar = britecharts.groupedBar();
let groupedBarData = [ 
   { group: "Unique Visits", name: "2011-01-05T00:00:00", value: 56 },
   { group: "Leads", name: "2011-01-05T00:00:00", value: 12 },
   { group: "Calls", name: "2011-01-05T00:00:00", value: 6 }];

const chartTooltip = britecharts.tooltip();
let containerWidth = container.node() ? container.node().getBoundingClientRect().width : false;
if (containerWidth) {
   groupedBar
      .tooltipThreshold(400)
      .width(containerWidth)
      .grid('horizontal')
      .isAnimated(true)
      .colorSchema(britecharts.colors.colorSchemas.britecharts)
      .on('customMouseOver', chartTooltip.show)
      .on('customMouseMove', function (dataPoint, topicColorMap, x, y) {
         chartTooltip.update(dataPoint, topicColorMap, x, y);
      })
      .on('customMouseOut', chartTooltip.hide);

   container.datum(groupedBarData).call(groupedBar);
   chartTooltip
      .topicLabel('values')
      .dateLabel('key')
      .title('Date');
   const tooltipContainer = d3.select('.bar-container .metadata-group');
   tooltipContainer.datum([]).call(chartTooltip);
}
```

**Comment by Golodhros** [comment]
Thanks for the extra details @SazanYa! This will help us trying to fix it.

Could you give it a try with the latest RC and see if it still happens? You can install it with `npm i britecharts@next`

**Comment by Golodhros** [comment]
Fixed in 3.0.0-alpha-6.1.2
